File: src/main/java/com/collectingwithzak/service/DashboardService.java
package com.collectingwithzak.service;

import com.collectingwithzak.dto.dashboard.DashboardData;
import com.collectingwithzak.dto.dashboard.MonthlyRevenue;
import com.collectingwithzak.dto.inventory.InventoryTotals;
import com.collectingwithzak.dto.sale.RangeTotals;
import com.collectingwithzak.entity.enums.ItemStatus;
import com.collectingwithzak.mapper.LotMapper;
import com.collectingwithzak.mapper.SaleMapper;
import com.collectingwithzak.repository.GradingSubmissionRepository;
import com.collectingwithzak.repository.LotPurchaseRepository;
import com.collectingwithzak.repository.SaleRepository;
import com.collectingwithzak.repository.TrackedItemRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.function.ToDoubleFunction;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class DashboardService {

    private static final int TIMELINE_MONTHS = 12;
    private static final int TOP_N = 5;
    private static final DateTimeFormatter MONTH_FMT = DateTimeFormatter.ofPattern("yyyy-MM");

    private final LotPurchaseRepository lotRepo;
    private final SaleRepository saleRepo;
    private final TrackedItemRepository itemRepo;
    private final GradingSubmissionRepository gradingRepo;
    private final SaleMapper saleMapper;
    private final LotMapper lotMapper;
    private final VincePaymentService vincePaymentService;

    public DashboardData getDashboardData() {
        RangeTotals confirmed = saleRepo.getConfirmedTotals();
        double totalSpent = lotRepo.getTotalCostNonRejected();
        InventoryTotals invTotals = itemRepo.getInventoryTotals();

        List<String> monthLabels = buildMonthLabels();
        Map<String, MonthlyRevenue> revenueByMonth = saleRepo.getMonthlyRevenue(TIMELINE_MONTHS).stream()
                .collect(Collectors.toMap(MonthlyRevenue::getMonth, Function.identity()));
        Map<String, Double> spendByMonth = lotRepo.getMonthlySpend(TIMELINE_MONTHS);

        return DashboardData.builder()
                .totalSpent(totalSpent)
                .totalGross(confirmed.getGross())
                .totalNet(confirmed.getNet())
                .totalFees(confirmed.getFees())
                .margin(confirmed.getNet() - totalSpent)
                .salesCount(confirmed.getCount())
                .avgSale(confirmed.getCount() > 0 ? confirmed.getNet() / confirmed.getCount() : 0)
                .gradingCount(itemRepo.countByStatus(ItemStatus.IN_GRADING.name()))
                .inventoryCount(itemRepo.countByStatus(ItemStatus.AVAILABLE.name()))
                .inventoryCost(invTotals.getCost())
                .inventoryMarket(invTotals.getMarket())
                .totals7(saleRepo.getTotalsSince(LocalDate.now().minusDays(7)))
                .totals30(saleRepo.getTotalsSince(LocalDate.now().minusDays(30)))
                .monthLabels(monthLabels)
                .monthlySpend(fillSeries(monthLabels, spendByMonth))
                .monthlyGross(fillRevenueSeries(monthLabels, revenueByMonth, MonthlyRevenue::getGross))
                .monthlyNet(fillRevenueSeries(monthLabels, revenueByMonth, MonthlyRevenue::getNet))
                .originCounts(saleRepo.countByOrigin())
                .itemTypeCounts(itemRepo.countByItemType())
                .gradingStatuses(gradingRepo.countByStatus())
                .lotStatuses(lotRepo.countByStatus())
                .topSales(saleMapper.toResponseList(saleRepo.findTopByNet(PageRequest.of(0, TOP_N))))
                .recentSales(saleMapper.toResponseList(saleRepo.findRecent(PageRequest.of(0, TOP_N))))
                .recentLots(lotMapper.toResponseList(lotRepo.findByOrderByPurchaseDateDesc(PageRequest.of(0, TOP_N))))
                .vinceLedger(vincePaymentService.getLedger())
                .build();
    }

    private List<String> buildMonthLabels() {
        List<String> labels = new ArrayList<>(TIMELINE_MONTHS);
        YearMonth current = YearMonth.now();
        for (int i = TIMELINE_MONTHS - 1; i >= 0; i--) {
            labels.add(current.minusMonths(i).format(MONTH_FMT));
        }
        return labels;
    }

    private List<Double> fillSeries(List<String> labels, Map<String, Double> data) {
        return labels.stream()
                .map(label -> data.getOrDefault(label, 0.0))
                .toList();
    }

    private List<Double> fillRevenueSeries(List<String> labels, Map<String, MonthlyRevenue> data,
                                            ToDoubleFunction<MonthlyRevenue> extractor) {
        return labels.stream()
                .map(label -> {
                    MonthlyRevenue revenue = data.get(label);
                    return revenue != null ? extractor.applyAsDouble(revenue) : 0.0;
                })
                .toList();
    }
}

File: go-app/internal/services/dashboard_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/zakpruitt/pbst/internal/models"
	"github.com/zakpruitt/pbst/internal/repository"
)

type DashboardService interface {
	GetDashboardData(ctx context.Context) (DashboardData, error)
}

type dashboardService struct {
	lotRepo     repository.LotRepository
	saleRepo    repository.SaleRepository
	itemRepo    repository.TrackedItemRepository
	gradingRepo repository.GradingRepository
}

func NewDashboardService(
	lotRepo repository.LotRepository,
	saleRepo repository.SaleRepository,
	itemRepo repository.TrackedItemRepository,
	gradingRepo repository.GradingRepository,
) DashboardService {
	return &dashboardService{
		lotRepo:     lotRepo,
		saleRepo:    saleRepo,
		itemRepo:    itemRepo,
		gradingRepo: gradingRepo,
	}
}

type DashboardData struct {
	Page            string
	TotalSpent      float64
	TotalGross      float64
	TotalNet        float64
	TotalFees       float64
	Margin          float64
	SalesCount      int64
	AvgSale         float64
	GradingCount    int64
	InventoryCount  int64
	InventoryCost   float64
	InventoryMarket float64
	Totals7         repository.RangeTotals
	Totals30        repository.RangeTotals
	MonthLabels     []string
	MonthlySpend    []float64
	MonthlyGross    []float64
	MonthlyNet      []float64
	OriginCounts    []repository.OriginCount
	ItemTypeCounts  []repository.ItemTypeCount
	GradingStatuses []repository.GradingStatusCount
	LotStatuses     []repository.LotStatusCount
	TopSales        []models.Sale
	RecentSales     []models.Sale
	RecentLots      []models.LotPurchase
	VinceTotals     repository.RangeTotals
}

func (s *dashboardService) GetDashboardData(ctx context.Context) (DashboardData, error) {
	data := DashboardData{Page: "dashboard"}

	totalSpent, err := s.lotRepo.GetTotalCostNonRejected(ctx)
	if err != nil {
		return data, fmt.Errorf("total spent: %w", err)
	}
	totalGross, err := s.saleRepo.GetTotalGrossAmount(ctx)
	if err != nil {
		return data, fmt.Errorf("total gross: %w", err)
	}
	totalNet, err := s.saleRepo.GetTotalNetAmount(ctx)
	if err != nil {
		return data, fmt.Errorf("total net: %w", err)
	}
	totalFees, err := s.saleRepo.GetTotalFees(ctx)
	if err != nil {
		return data, fmt.Errorf("total fees: %w", err)
	}

	gradingCount, err := s.itemRepo.CountByPurpose(ctx, "IN_GRADING")
	if err != nil {
		return data, fmt.Errorf("grading count: %w", err)
	}
	inventoryCount, err := s.itemRepo.CountByPurpose(ctx, "INVENTORY")
	if err != nil {
		return data, fmt.Errorf("inventory count: %w", err)
	}
	invCost, invMarket, err := s.itemRepo.InventoryTotals(ctx)
	if err != nil {
		return data, fmt.Errorf("inventory totals: %w", err)
	}

	salesCount, err := s.saleRepo.CountByStatus(ctx, "CONFIRMED")
	if err != nil {
		return data, fmt.Errorf("sales count: %w", err)
	}
	avgSale := 0.0
	if salesCount > 0 {
		avgSale = totalNet / float64(salesCount)
	}

	monthlyRev, err := s.saleRepo.MonthlyRevenue(ctx, 12)
	if err != nil {
		return data, fmt.Errorf("monthly revenue: %w", err)
	}
	monthlySpend, err := s.lotRepo.MonthlySpend(ctx, 12)
	if err != nil {
		return data, fmt.Errorf("monthly spend: %w", err)
	}

	originCounts, err := s.saleRepo.CountByOrigin(ctx)
	if err != nil {
		return data, fmt.Errorf("origin counts: %w", err)
	}
	itemTypeCounts, err := s.itemRepo.CountByItemType(ctx)
	if err != nil {
		return data, fmt.Errorf("item type counts: %w", err)
	}
	gradingStatuses, err := s.gradingRepo.CountByStatus(ctx)
	if err != nil {
		return data, fmt.Errorf("grading status counts: %w", err)
	}
	lotStatuses, err := s.lotRepo.CountByStatus(ctx)
	if err != nil {
		return data, fmt.Errorf("lot status counts: %w", err)
	}

	topSales, err := s.saleRepo.GetTopByNet(ctx, 5)
	if err != nil {
		return data, fmt.Errorf("top sales: %w", err)
	}
	recentSales, err := s.saleRepo.GetRecent(ctx, 5)
	if err != nil {
		return data, fmt.Errorf("recent sales: %w", err)
	}
	recentLots, err := s.lotRepo.GetRecent(ctx, 5)
	if err != nil {
		return data, fmt.Errorf("recent lots: %w", err)
	}

	totals30, err := s.saleRepo.TotalsSince(ctx, time.Now().AddDate(0, 0, -30))
	if err != nil {
		return data, fmt.Errorf("30d totals: %w", err)
	}
	totals7, err := s.saleRepo.TotalsSince(ctx, time.Now().AddDate(0, 0, -7))
	if err != nil {
		return data, fmt.Errorf("7d totals: %w", err)
	}
	vinceTotals, err := s.saleRepo.VinceTotals(ctx)
	if err != nil {
		return data, fmt.Errorf("vince totals: %w", err)
	}

	labels, spendSeries, grossSeries, netSeries := buildMonthlySeries(monthlyRev, monthlySpend, 12)

	data.TotalSpent = totalSpent
	data.TotalGross = totalGross
	data.TotalNet = totalNet
	data.TotalFees = totalFees
	data.Margin = totalNet - totalSpent
	data.SalesCount = salesCount
	data.AvgSale = avgSale
	data.GradingCount = gradingCount
	data.InventoryCount = inventoryCount
	data.InventoryCost = invCost
	data.InventoryMarket = invMarket
	data.Totals7 = totals7
	data.Totals30 = totals30
	data.MonthLabels = labels
	data.MonthlySpend = spendSeries
	data.MonthlyGross = grossSeries
	data.MonthlyNet = netSeries
	data.OriginCounts = originCounts
	data.ItemTypeCounts = itemTypeCounts
	data.GradingStatuses = gradingStatuses
	data.LotStatuses = lotStatuses
	data.TopSales = topSales
	data.RecentSales = recentSales
	data.RecentLots = recentLots
	data.VinceTotals = vinceTotals
	return data, nil
}

// buildMonthlySeries zero-fills a 12-month timeline so the P&L chart has
// aligned x-axis labels even for months with only purchases or only sales.
func buildMonthlySeries(rev []repository.MonthlyRevenue, spend []repository.MonthlySpend, months int) (labels []string, spendSeries, grossSeries, netSeries []float64) {
	labels = make([]string, 0, months)
	now := time.Now()
	for i := months - 1; i >= 0; i-- {
		t := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC).AddDate(0, -i, 0)
		labels = append(labels, t.Format("2006-01"))
	}

	revBy := make(map[string]repository.MonthlyRevenue, len(rev))
	for _, r := range rev {
		revBy[r.Month] = r
	}
	spendBy := make(map[string]repository.MonthlySpend, len(spend))
	for _, sp := range spend {
		spendBy[sp.Month] = sp
	}

	spendSeries = make([]float64, len(labels))
	grossSeries = make([]float64, len(labels))
	netSeries = make([]float64, len(labels))
	for i, m := range labels {
		spendSeries[i] = spendBy[m].Spend
		grossSeries[i] = revBy[m].Gross
		netSeries[i] = revBy[m].Net
	}
	return labels, spendSeries, grossSeries, netSeries
}
